File: Matlab/utils/cal_Stokes.m
function [S0,DoLP,AoLP] = cal_Stokes(im)
imOut0 = im(:,:,1:3);
imOut45 = im(:,:,4:6);
imOut90 = im(:,:,7:9);
imOut135 = im(:,:,10:12);

S0 = (imOut0+imOut45+imOut90+imOut135)*0.5;
S1 = imOut0-imOut90;
S2 = imOut45-imOut135;
DoLP = sqrt(S1.^2+S2.^2)./(S0+eps);
AoLP = 0.5*(atan2(S2, S1));
% angles = 0.5*(atan2(S2, S1));
% AoLP = rad2deg(angles);
% AoLP(AoLP < 0) = AoLP(AoLP < 0) + 360;
end

File: Python/utils/cal_Stokes.py
import numpy as np


def cal_Stokes(im):
    im_out0 = im[:, :, 0:3]
    im_out45 = im[:, :, 3:6]
    im_out90 = im[:, :, 6:9]
    im_out135 = im[:, :, 9:12]

    s0 = (im_out0 + im_out45 + im_out90 + im_out135) * 0.5
    s1 = im_out0 - im_out90
    s2 = im_out45 - im_out135
    dolp = np.sqrt(s1 ** 2 + s2 ** 2) / (s0 + np.finfo(np.float64).eps)
    aolp = 0.5 * np.arctan2(s2, s1)

    return s0, dolp, aolp
